homeassistant/components/ozw/config_flow.py

```diff
@@ -1,24 +1,174 @@
+        """Set up flow instance."""
+        self.addon_config = None
+        self.addon_info = None
+        self.network_key = None
+        self.usb_path = None
+        # If we install the the add-on we should uninstall it on entry unload.
+        self.integration_created_addon = False
+
     async def async_step_user(self, user_input=None):
         """Handle the initial step."""
         if self._async_current_entries():
             return self.async_abort(reason="single_instance_allowed")
+
+        if not self.hass.components.hassio.is_hassio():
+            return self._async_use_mqtt_integration()
+
+        return await self.async_step_on_supervisor()
+
+    @callback
+    def _async_use_mqtt_integration(self):
+        """Handle logic when using the MQTT integration."""
         if "mqtt" not in self.hass.config.components:
             return self.async_abort(reason="mqtt_required")
+
+        return self.async_create_entry(
+            title=TITLE,
+            data={
+                CONF_USB_PATH: self.usb_path,
+                CONF_NETWORK_KEY: self.network_key,
+                CONF_INTEGRATION_CREATED_ADDON: self.integration_created_addon,
+            },
+        )
+
+    async def async_step_on_supervisor(self, user_input=None):
+        """Handle logic when on Supervisor host."""
+        if user_input is not None:
+            if user_input[CONF_USE_ADDON]:
+                return await self._async_use_addon()
+            return self._async_use_mqtt_integration()
+
+        return self.async_show_form(
+            step_id="on_supervisor", data_schema=ON_SUPERVISOR_SCHEMA
+        )
+
+    async def _async_use_addon(self):
+        """Handle logic when using the OpenZWave add-on."""
+        if await self._async_is_addon_running():
+            return self._async_use_mqtt_integration()
+
+        if await self._async_is_addon_installed():
+            return await self.async_step_start_addon()
+
+        return await self._async_install_addon()
+
+    async def _async_install_addon(self):
+        """Install add-on."""
+        try:
+            await self.hass.components.hassio.async_install_addon("core_zwave")
+        except HassioAPIError as err:
+            _LOGGER.error("Failed to install OpenZWave add-on: %s", err)
+            return self.async_abort(reason="addon_install_failed")
+        self.integration_created_addon = True
+
+        return await self.async_step_start_addon()
+
+    async def async_step_start_addon(self, user_input=None):
+        """Ask for missing config and start add-on."""
+        if self.addon_config is None:
+            self.addon_config = await self._async_get_addon_config()
+
+        self.usb_path = self.addon_config.get(CONF_ADDON_DEVICE)
+        self.network_key = self.addon_config.get(CONF_ADDON_NETWORK_KEY)
+
+        if user_input is None and None in (self.usb_path, self.network_key):
+            data_schema = vol.Schema({})
+
+            if not self.usb_path:
```

Comment:
Why would we not allow configuring USB path if it's already configured? Shouldn't we just always allow changing it here ?

Reply:
Yeah, I agree. I was following what we had written but it's better to ask for everything I think.